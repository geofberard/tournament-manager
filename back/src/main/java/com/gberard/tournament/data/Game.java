package com.gberard.tournament.data;

import lombok.Builder;
import org.apache.commons.codec.digest.DigestUtils;

import java.time.LocalDateTime;
import java.util.Optional;
import java.util.OptionalInt;

import static com.gberard.tournament.data.GameTeamStatus.*;

public record Game(
        String id,
        LocalDateTime time,
        String court,
        Team teamA,
        Team teamB,
        Optional<Team> referee,
        Optional<Integer> scoreA,
        Optional<Integer> scoreB
) {
    public boolean isFinished () {
        return scoreA.isPresent() && scoreB.isPresent();
    }

    public GameTeamStatus getTeamStatus(Team team) {
        if (!hasContestant(team)) {
            throw new IllegalStateException("Team " + team.id() + " has to played game " + this.id);
        }

        if (!isFinished()) {
            return NOT_PLAYED;
        }

        if (scoreA.get() == scoreB.get()) {
            return DRAWN;
        }

        return getPointsFor(team).get() > getPointsAgainst(team).get() ? WIN : LOST;
    }

    public boolean hasContestant(Team team) {
        return teamA.equals(team) || teamB.equals(team);
    }

    public Optional<Integer> getPointsFor(Team team) {
        if (!hasContestant(team)) {
            throw new IllegalStateException("Team " + team.id() + " has to played game " + this.id);
        }
        return (teamA.equals(team) ? scoreA : scoreB);
    }

    public Optional<Integer> getPointsAgainst(Team team) {
        if (!hasContestant(team)) {
            throw new IllegalStateException("Team " + team.id() + "has to played game " + this.id);
        }
        return (teamA.equals(team) ? scoreB : scoreA);
    }

    @Builder(builderMethodName = "builder")
    public static Game createGame(
            LocalDateTime time,
            String court,
            Team teamA,
            Team teamB,
            Team referee,
            Integer scoreA,
            Integer scoreB
    ) {
        return new Game(DigestUtils.sha1Hex(court + time.toString()), time, court, teamA, teamB,
                Optional.ofNullable(referee), Optional.ofNullable(scoreA), Optional.ofNullable(scoreB));
    }
}