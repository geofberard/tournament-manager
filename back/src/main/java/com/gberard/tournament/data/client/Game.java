package com.gberard.tournament.data.client;

import com.gberard.tournament.data.score.Score;
import com.gberard.tournament.data.score.ScoreType;
import lombok.Builder;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

public record Game(
        String id,
        LocalDateTime time,
        String court,
        List<String> contestantIds,
        Optional<String> refereeId,
        Boolean isFinished,
        ScoreType scoreType,
        Optional<Score> score
) implements Identified {

    @Builder
    public static Game createGame(
            String id,
            LocalDateTime time,
            String court,
            List<String> contestantIds,
            String refereeId,
            Boolean isFinished,
            ScoreType scoreType,
            Score score
    ) {
        return new Game(id, time, court, contestantIds, Optional.ofNullable(refereeId), isFinished, scoreType,
                Optional.ofNullable(score));
    }
}