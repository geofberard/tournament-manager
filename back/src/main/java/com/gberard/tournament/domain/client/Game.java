package com.gberard.tournament.domain.client;

import com.gberard.tournament.domain.score.Score;
import com.gberard.tournament.domain.score.ScoreType;
import lombok.Builder;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

public record Game(
        String id,
        LocalDateTime time,
        String court,
        List<String> contestantIds,
        Optional<String> refereeId,
        Boolean isFinished,
        ScoreType scoreType,
        Optional<Score> score
) implements Identified {

    @Builder
    public static Game createGame(
            String id,
            LocalDateTime time,
            String court,
            List<String> contestantIds,
            String refereeId,
            Boolean isFinished,
            ScoreType scoreType,
            Score score
    ) {
        return new Game(id, time, court, contestantIds, Optional.ofNullable(refereeId), isFinished, scoreType,
                Optional.ofNullable(score));
    }
}