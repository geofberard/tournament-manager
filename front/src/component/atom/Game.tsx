import { SxProps, TableRow, TableCell, tableCellClasses, styled } from "@mui/material";
import * as React from "react";
import { FC } from "react";

// const menuContainer: SxProps = {
//     backgroundColor: "primary.dark",
//     color: "primary.contrastText",
// };
const StyledTableCell = styled(TableCell)(({ theme }) => ({
    // [`&.${tableCellClasses.head}`]: {
    //     backgroundColor: theme.palette.common.black,
    //     color: theme.palette.common.white,
    // },
    [`&.${tableCellClasses.body}`]: {
        // fontSize: 14,
        color: theme.palette.common.white,
    },
}));

const StyledTableRow = styled(TableRow)(({ theme }) => ({
    '&:nth-of-type(odd)': {
        backgroundColor: theme.palette.action.hover,
    },
}));

export const Game: FC = () => (
    <StyledTableRow
        key={1}
        sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
    >
        <StyledTableCell align="center" component="th" scope="row">FT</StyledTableCell>
        <StyledTableCell align="center">Team a</StyledTableCell>
        <StyledTableCell align="center">3 - 0</StyledTableCell>
        <StyledTableCell align="center">Team b</StyledTableCell>
    </StyledTableRow>
);

