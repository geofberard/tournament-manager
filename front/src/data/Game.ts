import { Contestant } from "./Contestant";
import { Identified } from "./Identified";
import { ScoreType } from "./ScoreType";
import { Score } from "./Score";
import {Team} from "./Team";

export enum Result {
    WON = "WON",
    LOST = "LOST",
    DEUCE = "DEUCE",
    NOT_PLAYED = "NOT_PLAYED",
}

export default class Game implements Identified {

    id: string;
    time: Date;
    court: string;
    contestants: Contestant[];
    referee?: Contestant;
    isFinished: boolean;
    scoreType: ScoreType;
    score?: Score;


    public constructor(init?:Partial<Game>) {
        Object.assign(this, init);
    }

    // isFinished: () => boolean = () => !(this.scoreB == undefined || this.scoreA == undefined);

    getResultGameByTeam = (contestant: Contestant): Result => {
        if(!this.isFinished){
            return Result.NOT_PLAYED;
        }
        return Result.NOT_PLAYED;

        // @Patator : to be reimplemented depending on the type of score
        // if(this.scoreA === this.scoreB) {
        //     return Result.DEUCE;
        // }
        //
        // if(this.scoreA > this.scoreB) {
        //     return (this.teamA === team) ? Result.WON : Result.LOST
        // }
        //
        // return (this.teamB === team) ? Result.WON : Result.LOST;
    };

}

