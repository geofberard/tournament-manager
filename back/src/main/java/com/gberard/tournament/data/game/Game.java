package com.gberard.tournament.data.game;

import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.gberard.tournament.data.Identified;
import com.gberard.tournament.data.contestant.Contestant;
import com.gberard.tournament.data.game.score.Score;
import lombok.Builder;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

import static com.gberard.tournament.data.game.ContestantResult.*;

public record Game(
        String id,
        LocalDateTime time,
        String court,
        List<Contestant> contestants,
        Optional<Contestant> referee,
        Optional<Score> score
) implements Identified {

    public boolean isFinished() {
        return score.isPresent();
    }

    public boolean hasContestant(Contestant contestant) {
        return contestants.contains(contestant);
    }

    @Builder
    public static Game createGame(
            String id,
            LocalDateTime time,
            String court,
            List<Contestant> contestants,
            Contestant referee,
            Score score
    ) {
        return new Game(id, time, court, contestants, Optional.ofNullable(referee), Optional.ofNullable(score));
    }
}