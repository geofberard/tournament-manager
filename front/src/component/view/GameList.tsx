import { Grid, SxProps, TableContainer, Table, TableBody, Typography, Theme} from "@mui/material";
import * as React from "react";
import { FC } from "react";
import Game from "../../data/Game";
import { RowGame } from "../atom/RowGame";
import { useTeams } from "../hook/useTeams";

const gameContainer: SxProps = {
    backgroundColor: "primary.dark",
    color: "primary.contrastText",
    paddingLeft: { xs: "0", md: "15px" }, 
    paddingRight: { xs: "0", md: "15px" }, 
};

const titleContainer: SxProps = {
    maxWidth: "300px",
    margin: "0 auto",
    marginBottom: "5px",
    paddingTop: "10px" ,
    paddingBottom: { xs: "0", md: "10px" },
    borderBottom: (theme: Theme) => `5px solid ${theme.palette.secondary.main}`,
};

export const GameList: FC = () => {
    const teams = useTeams();

    const games = [
        new Game("Game 1", new Date("2022-01-01 11:00") , "Terrain 1", teams[0], teams[1], undefined, 25, 7),
        new Game("Game 2", new Date("2022-01-01 12:00"), "Terrain 3", teams[0], teams[2], undefined, 12, 25),
        new Game("Game 3", new Date("2022-01-01 13:00"), "Terrain 2", teams[0], teams[3], undefined, 22, 22),
        new Game("Game 4", new Date("2022-01-01 14:00"), "Terrain 4", teams[0], teams[4]),
        new Game("Game 5", new Date("2022-01-01 15:00"), "Terrain 1", teams[0], teams[5]),
        new Game("Game 6", new Date("2022-01-01 16:00"), "Terrain 2", teams[0], teams[6]),
    ];

    return (
        <Grid
            container
            item
            md={5}
            direction={{ xs: "row", md: "column" }}
            sx={gameContainer}
        >
            <Grid item sx={titleContainer}>
                <Typography variant="h3">Vos Matchs</Typography>
            </Grid>
            <TableContainer >
                <Table>
                    <TableBody>
                        {games.map((game, index) => (
                            <RowGame key={index} game={game} />
                        ))}
                    </TableBody>
                </Table>
            </TableContainer>
        </Grid>
    );
};
