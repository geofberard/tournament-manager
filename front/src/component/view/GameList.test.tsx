import * as React from "react";
import { testSnapshot } from "../../test/TestUtils";
import { GameList } from "./GameList";
import Game from "../../data/Game";

const teamA = { id: "teamA", name: "TeamA" };
const teamB = { id: "teamB", name: "TeamB" };
const teamC = { id: "teamC", name: "TeamC" };

const games: Game[] = [
    new Game({
        id: "gameId",
        time: new Date("2022-08-19 15:00"),
        court: "court",
        teamA: teamA,
        teamB: teamB,
        referee: teamC,
        scoreA: 25,
        scoreB: 14,
    }),
    new Game({
        id: "gameId",
        time: new Date("2022-08-19 16:00"),
        court: "court",
        teamA: teamA,
        teamB: teamB,
        referee: teamC,
        scoreA: 14,
        scoreB: 25,
    }),
    new Game({
        id: "gameId",
        time: new Date("2022-08-19 17:00"),
        court: "court",
        teamA: teamA,
        teamB: teamB,
        referee: teamC,
        scoreA: 22,
        scoreB: 22,
    })
    
];


testSnapshot(<GameList title="test" games={games} />);
