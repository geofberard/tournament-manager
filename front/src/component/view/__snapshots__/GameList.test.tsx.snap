// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match last snapshot 1`] = `
<ForwardRef(Grid)
  container={true}
  direction={
    Object {
      "md": "column",
      "xs": "row",
    }
  }
  item={true}
  md={5}
  sx={
    Object {
      "backgroundColor": "primary.dark",
      "color": "primary.contrastText",
      "paddingLeft": Object {
        "md": "15px",
        "xs": "0",
      },
      "paddingRight": Object {
        "md": "15px",
        "xs": "0",
      },
    }
  }
>
  <ForwardRef(Grid)
    item={true}
    sx={
      Object {
        "borderBottom": [Function],
        "margin": "0 auto",
        "marginBottom": "5px",
        "maxWidth": "300px",
        "paddingBottom": Object {
          "md": "10px",
          "xs": "0",
        },
        "paddingTop": "10px",
      }
    }
  >
    <ForwardRef(Typography)
      variant="h3"
    >
      test
    </ForwardRef(Typography)>
  </ForwardRef(Grid)>
  <GameTable
    games={
      Array [
        Game {
          "court": "court",
          "getResultGameByTeam": [Function],
          "id": "gameId",
          "isFinished": [Function],
          "referee": Object {
            "id": "teamC",
            "name": "TeamC",
          },
          "scoreA": 25,
          "scoreB": 14,
          "teamA": Object {
            "id": "teamA",
            "name": "TeamA",
          },
          "teamB": Object {
            "id": "teamB",
            "name": "TeamB",
          },
          "time": 2022-08-19T13:00:00.000Z,
        },
        Game {
          "court": "court",
          "getResultGameByTeam": [Function],
          "id": "gameId",
          "isFinished": [Function],
          "referee": Object {
            "id": "teamC",
            "name": "TeamC",
          },
          "scoreA": 14,
          "scoreB": 25,
          "teamA": Object {
            "id": "teamA",
            "name": "TeamA",
          },
          "teamB": Object {
            "id": "teamB",
            "name": "TeamB",
          },
          "time": 2022-08-19T14:00:00.000Z,
        },
        Game {
          "court": "court",
          "getResultGameByTeam": [Function],
          "id": "gameId",
          "isFinished": [Function],
          "referee": Object {
            "id": "teamC",
            "name": "TeamC",
          },
          "scoreA": 22,
          "scoreB": 22,
          "teamA": Object {
            "id": "teamA",
            "name": "TeamA",
          },
          "teamB": Object {
            "id": "teamB",
            "name": "TeamB",
          },
          "time": 2022-08-19T15:00:00.000Z,
        },
      ]
    }
  />
</ForwardRef(Grid)>
`;
